src/game/BattleGround/BattleGroundWS.cpp

```diff
@@ -217,218 +293,143 @@ void BattleGroundWS::EventPlayerCapturedFlag(Player* source)
 
-        m_FlagsDropTimer[GetOtherTeamIndex(GetTeamIndexByTeamId(source->GetTeam()))] = BG_WS_FLAG_DROP_TIME;
+        SendMessageToAll(wsMessageIds[otherTeamIdx][BG_WS_FLAG_ACTION_DROPPED],
+                         wsChatMessageTypes[TEAM_INDEX_HORDE][BG_WS_FLAG_ACTION_DROPPED], player);
+        UpdateWorldState(wsStateUpdateId[otherTeamIdx], uint32(-1));
+        m_FlagsDropTimer[otherTeamIdx] = BG_WS_FLAG_DROP_TIME;
     }
 }
 
-void BattleGroundWS::EventPlayerClickedOnFlag(Player* source, GameObject* target_obj)
+void BattleGroundWS::PickUpFlagFromBase(Player* player)
 {
-    if (GetStatus() != STATUS_IN_PROGRESS)
-        return;
+    PvpTeamIndex teamIdx = GetTeamIndexByTeamId(player->GetTeam());
+    PvpTeamIndex otherTeamIdx = GetOtherTeamIndex(teamIdx);
+
+    PlaySoundToAll(wsSounds[otherTeamIdx][BG_WS_FLAG_ACTION_PICKEDUP]);
+    SpawnEvent(otherTeamIdx, 0, false);
+    SetFlagCarrier(otherTeamIdx, player->GetObjectGuid());
+    m_FlagState[otherTeamIdx] = BG_WS_FLAG_STATE_ON_PLAYER;
+
+    // update world state to show correct flag carrier
+    UpdateFlagState(player->GetTeam(), BG_WS_FLAG_STATE_ON_PLAYER);
+    UpdateWorldState(wsStateUpdateId[otherTeamIdx], 1);
+
+    player->CastSpell(player, wsSpellTypes[otherTeamIdx][BG_WS_FLAG_ACTION_PICKEDUP], TRIGGERED_OLD_TRIGGERED);
+    SendMessageToAll(wsMessageIds[otherTeamIdx][BG_WS_FLAG_ACTION_PICKEDUP],
+                     wsChatMessageTypes[teamIdx][BG_WS_FLAG_ACTION_PICKEDUP], player);
+    player->RemoveAurasWithInterruptFlags(AURA_INTERRUPT_FLAG_ENTER_PVP_COMBAT);
+}
 
-    int32 message_id = 0;
-    ChatMsg type = CHAT_MSG_BG_SYSTEM_NEUTRAL;
+uint32 BattleGroundWS::GroundFlagInteraction(Player* player, GameObject* target)
+{
+    PvpTeamIndex teamIdx = GetTeamIndexByTeamId(player->GetTeam());
+    PvpTeamIndex otherTeamIdx = GetOtherTeamIndex(teamIdx);
 
-    uint8 event = (sBattleGroundMgr.GetGameObjectEventIndex(target_obj->GetGUIDLow())).event1;
+    int32 actionId = -1;
+    uint32 displayId = target->GetDisplayId();
 
-    // alliance flag picked up from base
-    if (source->GetTeam() == HORDE && GetFlagState(ALLIANCE) == BG_WS_FLAG_STATE_ON_BASE
-            && event == WS_EVENT_FLAG_A)
+    // check if we are returning our flag
+    if (wsFlagIds[teamIdx] == displayId)
     {
-        message_id = LANG_BG_WS_PICKEDUP_AF;
-        type = CHAT_MSG_BG_SYSTEM_HORDE;
-        PlaySoundToAll(BG_WS_SOUND_ALLIANCE_FLAG_PICKED_UP);
-        SpawnEvent(WS_EVENT_FLAG_A, 0, false);
-        SetAllianceFlagCarrier(source->GetObjectGuid());
-        m_FlagState[TEAM_INDEX_ALLIANCE] = BG_WS_FLAG_STATE_ON_PLAYER;
-        // update world state to show correct flag carrier
-        UpdateFlagState(HORDE, BG_WS_FLAG_STATE_ON_PLAYER);
-        UpdateWorldState(BG_WS_FLAG_UNK_ALLIANCE, 1);
-        source->CastSpell(source, BG_WS_SPELL_SILVERWING_FLAG, TRIGGERED_OLD_TRIGGERED);
+        actionId = BG_WS_FLAG_ACTION_RETURNED;
+        Team team = player->GetTeam();
```

Comment:
GetTeam() should be cached at start of function and used in all cases instead of calling the function